File: bench/algorithm/regex-redux/6.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// contributed by Tom Kaitchuck
// removed parallelization

use std::io::{self, BufWriter, Read, Write};
use std::mem;

struct Regex {
    string: &'static str,
    regex: ::regex::bytes::Regex,
}

impl Regex {
    fn new(string: &'static str) -> Regex {
        Regex {
            string: string,
            regex: ::regex::bytes::Regex::new(string).unwrap(),
        }
    }

    fn replace_all<'t>(&self, text: &'t [u8], rep: &[u8], out: &mut Vec<u8>) {
        let mut last_match = 0;
        for m in self.regex.find_iter(text) {
            out.extend_from_slice(&text[last_match..m.start()]);
            out.extend_from_slice(&rep);
            last_match = m.end();
        }
        out.extend_from_slice(&text[last_match..]);
    }
}

fn count_reverse_complements(sequence: &Vec<u8>) -> Vec<String> {
    // Search for occurrences of the following patterns:
    let variants = vec![
        Regex::new("agggtaaa|tttaccct"),
        Regex::new("[cgt]gggtaaa|tttaccc[acg]"),
        Regex::new("a[act]ggtaaa|tttacc[agt]t"),
        Regex::new("ag[act]gtaaa|tttac[agt]ct"),
        Regex::new("agg[act]taaa|ttta[agt]cct"),
        Regex::new("aggg[acg]aaa|ttt[cgt]ccct"),
        Regex::new("agggt[cgt]aa|tt[acg]accct"),
        Regex::new("agggta[cgt]a|t[acg]taccct"),
        Regex::new("agggtaa[cgt]|[acg]ttaccct"),
    ];
    variants
        .iter()
        .map(|variant| {
            format!(
                "{} {}",
                variant.string,
                variant.regex.find_iter(sequence).count()
            )
        })
        .collect()
}

fn find_replaced_sequence_length(sequence: Vec<u8>, scratch_buff: Vec<u8>) -> usize {
    // Replace the following patterns, one at a time:
    let substs = vec![
        (Regex::new("tHa[Nt]"), &b"<4>"[..]),
        (Regex::new("aND|caN|Ha[DS]|WaS"), &b"<3>"[..]),
        (Regex::new("a[NSt]|BY"), &b"<2>"[..]),
        (Regex::new("<[^>]*>"), &b"|"[..]),
        (Regex::new("\\|[^|][^|]*\\|"), &b"-"[..]),
    ];

    let mut current = sequence;
    let mut next = scratch_buff;
    // Perform the replacements in sequence:
    for (re, replacement) in substs {
        re.replace_all(&current, replacement, &mut next);
        mem::swap(&mut current, &mut next);
        next.clear();
    }
    current.len()
}

fn main() -> anyhow::Result<()> {
    let file_name = std::env::args_os()
        .nth(1)
        .and_then(|s| s.into_string().ok())
        .unwrap_or("25000_in".into());

    let mut input = {
        let mut v = Vec::with_capacity(51 * (1 << 20));
        let mut file = std::fs::File::open(file_name).unwrap();
        file.read_to_end(&mut v).unwrap();
        v
    };
    let input_len = input.len();
    let mut sequence: Vec<u8> = Vec::with_capacity(input.len());
    Regex::new(">[^\n]*\n|\n").replace_all(&input, &b""[..], &mut sequence);
    let sequence_len = sequence.len();
    input.clear();
    let (result, counts) = rayon::join(
        || find_replaced_sequence_length(sequence.clone(), input),
        || count_reverse_complements(&sequence),
    );
    let mut stdout = BufWriter::new(io::stdout());
    for variant in counts {
        stdout.write_fmt(format_args!("{}\n", variant))?;
    }
    stdout.write_fmt(format_args!(
        "\n{}\n{}\n{:?}\n",
        input_len, sequence_len, result
    ))?;
    Ok(())
}

File: bench/algorithm/regex-redux/6.cs
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 
   Regex-Redux by Josh Goldfoot
   order variants by execution time by Anthony Lloyd
*/

using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

public static partial class RegexRedux
{
    [GeneratedRegex(@">.*\n|\n")]
    public static partial Regex ReplaceRegex();
    [GeneratedRegex(@"tHa[Nt]")]
    public static partial Regex MagicRe1();
    [GeneratedRegex(@"aND|caN|Ha[DS]|WaS")]
    public static partial Regex MagicRe2();
    [GeneratedRegex(@"a[NSt]|BY")]
    public static partial Regex MagicRe3();
    [GeneratedRegex(@"<[^>]*>")]
    public static partial Regex MagicRe4();
    [GeneratedRegex(@"\|[^|][^|]*\|")]
    public static partial Regex MagicRe5();
    [GeneratedRegex(@"[cgt]gggtaaa|tttaccc[acg]")]
    public static partial Regex Re2();
    [GeneratedRegex(@"a[act]ggtaaa|tttacc[agt]t")]
    public static partial Regex Re3();
    [GeneratedRegex(@"agggt[cgt]aa|tt[acg]accct")]
    public static partial Regex Re7();
    [GeneratedRegex(@"aggg[acg]aaa|ttt[cgt]ccct")]
    public static partial Regex Re6();
    [GeneratedRegex(@"ag[act]gtaaa|tttac[agt]ct")]
    public static partial Regex Re4();
    [GeneratedRegex(@"agg[act]taaa|ttta[agt]cct")]
    public static partial Regex Re5();
    [GeneratedRegex(@"agggtaaa|tttaccct")]
    public static partial Regex Re1();
    [GeneratedRegex(@"agggtaa[cgt]|[acg]ttaccct")]
    public static partial Regex Re9();
    [GeneratedRegex(@"agggta[cgt]a|t[acg]taccct")]
    public static partial Regex Re8();

    public static async Task Main(string[] args)
    {
        var fileName = args.Length > 0 ? args[0] : "25000_in";
        var sequences = await File.ReadAllTextAsync(fileName).ConfigureAwait(false);
        var initialLength = sequences.Length;
        sequences = ReplaceRegex().Replace(sequences, "");

        var magicTask = () =>
        {
            var newseq = MagicRe1().Replace(sequences, "<4>");
            newseq = MagicRe2().Replace(newseq, "<3>");
            newseq = MagicRe3().Replace(newseq, "<2>");
            newseq = MagicRe4().Replace(newseq, "|");
            newseq = MagicRe5().Replace(newseq, "-");
            return newseq.Length;
        };

        var variant2 = "[cgt]gggtaaa|tttaccc[acg] " + Re2().Count(_sequences);
        var variant3 = "a[act]ggtaaa|tttacc[agt]t " + Re3().Count(_sequences);
        var variant7 = "agggt[cgt]aa|tt[acg]accct " + Re7().Count(_sequences);
        var variant6 = "aggg[acg]aaa|ttt[cgt]ccct " + Re6().Count(_sequences);
        var variant4 = "ag[act]gtaaa|tttac[agt]ct " + Re4().Count(_sequences);
        var variant5 = "agg[act]taaa|ttta[agt]cct " + Re5().Count(_sequences);
        var variant1 = "agggtaaa|tttaccct " + Re1().Count(_sequences);
        var variant9 = "agggtaa[cgt]|[acg]ttaccct " + Re9().Count(_sequences);
        var variant8 = "agggta[cgt]a|t[acg]taccct " + Re8().Count(_sequences);

        Console.WriteLine(variant1);
        Console.WriteLine(variant2);
        Console.WriteLine(variant3);
        Console.WriteLine(variant4);
        Console.WriteLine(variant5);
        Console.WriteLine(variant6);
        Console.WriteLine(variant7);
        Console.WriteLine(variant8);
        Console.WriteLine(variant9);
        Console.WriteLine($"\n{initialLength}\n{sequences.Length}");
        Console.WriteLine(magicTask().ToString());
    }
}
